lib/sprockets/redirect.rb

```diff
@@ -74,14 +76,36 @@ def digest_path
 
     # Sends a redirect header back to browser
     def redirect_to_digest_version(env)
-      url = URI(@asset_host || @request.url)
-      url.path = "#{@prefix}/#{digest_path}"
+      url = File.join(compute_asset_host, @prefix, digest_path)
 
       headers = { 'Location'      => url.to_s,
                   'Content-Type'  => Rack::Mime.mime_type(::File.extname(digest_path)),
                   'Pragma'        => 'no-cache',
                   'Cache-Control' => 'no-cache; max-age=0' }
-      [self.class.redirect_status, headers, [redirect_message(url.to_s)]]
+
+      [self.class.redirect_status, headers, [redirect_message(url)]]
+    end
+
+    def compute_asset_host
+      host = @asset_host
+
+      if host.respond_to?(:call)
+        arity = host.respond_to?(:arity) ? host.arity : host.method(:call).arity
+        args = [logical_path]
+        args << @request if @request && (arity > 1 || arity < 0)
+        host = host.call(*args)
+      elsif host =~ /%d/
+        host = host % (Zlib.crc32(source) % 4)
+      end
+
+      host ||= @request.base_url if @request
+      return unless host
```

Comment:
I think the cause below can be removed altogether.

If user return something from Proc that's not a URL with scheme it's pretty much their bug and we shouldn't try to safeguard them.

I also feel like maybe this whole method should return an URI object so that you can just call `join` on line 79.